analyses/cell-type-wilms-tumor-06/notebook_template/utils/label-transfer-functions.R

```diff
@@ -0,0 +1,138 @@
+  )
+
+  # Calculate nCount_RNA and nFeature_RNA if the query does not
+  # contain them already
+  # Source: https://github.com/satijalab/azimuth/blob/243ee5db80fcbffa3452c944254a325a3da2ef9e/R/azimuth.R#L106-L120
+  if (!all(c("nCount_RNA", "nFeature_RNA") %in% c(colnames(x = query[[]])))) {
+    calcn <- as.data.frame(x = Seurat:::CalcN(object = query[["RNA"]]))
+    colnames(x = calcn) <- paste(
+      colnames(x = calcn),
+      NULL, # assay
+      sep = "_"
+    )
+    query <- AddMetaData(
+      object = query,
+      metadata = calcn
+    )
+    rm(calcn)
+  }
+
+  # Calculate percent mitochondrial genes if the query contains genes
+  # matching the regular expression "^MT-"
+  # Source: https://github.com/satijalab/azimuth/blob/243ee5db80fcbffa3452c944254a325a3da2ef9e/R/azimuth.R#L122-L131
+  if (any(grepl(pattern = "^MT-", x = rownames(x = query)))) {
+    query_symbols <- PercentageFeatureSet(
+      object = query,
+      pattern = "^MT-",
+      col.name = "percent.mt",
+      assay = NULL
+    )
+  }
+
+  return(query)
+}
+
+#' Perform label transfer using an Azimuth-adapted approach
+#'
+#' This function adapts code from the RunAzimuth function to perform label transfer:
+#' https://github.com/satijalab/azimuth/blob/243ee5db80fcbffa3452c944254a325a3da2ef9e/R/azimuth.R
+#'
+#' Note that several params were documented below based on their descriptions here:
+#' https://github.com/satijalab/seurat/blob/1549dcb3075eaeac01c925c4b4bb73c73450fc50/R/integration.R
+#'
+#' @param query The Seurat object which will undergo label transfer
+#' @param reference The reference dataset
+#' @param reference_dims Dimensions calculated from the reference dataset
+#' @param refdata object used by Azimuth
+#' @param reference_dims Number of dimensions to use in the anchor weighting procedure
+#' @param k.weight Number of neighbors to consider when weighting anchors. This should be
+#' <=15 when running on OpenScPCA test data
+#' @param n.trees More trees gives higher precision when using annoy approximate
+#' nearest neighbor search
+#' @param mapping.score.k Compute and store nearest k query neighbors in the
+#' AnchorSet object that is returned.
+#' @param ksmooth Number of cells to average over when computing transition
+#' probabilities
+#' @param verbose Display messages/progress
+#'
+#' @return Seurat object containing metadata from label transfer
+transfer_labels <- function(
+    query,
+    reference,
+    reference_dims,
+    refdata,
+    k.weight = 50,
+    n.trees = 20,
+    mapping.score.k = 80,
+    ksmooth = 80,
+    verbose = FALSE) {
+  # Find anchors between query and reference
+  # Source: https://github.com/satijalab/azimuth/blob/243ee5db80fcbffa3452c944254a325a3da2ef9e/R/azimuth.R#L132-L147
+  anchors <- FindTransferAnchors(
+    reference = reference,
+    query = query,
+    k.filter = NA,
+    reference.neighbors = "refdr.annoy.neighbors",
+    reference.assay = "refAssay",
+    query.assay = NULL,
```

Comment:
The `NULL` here is based on the source code, and to be honest I was hesitant to rock the boat on it...

This is defined in Seurat with [this `assay` argument](https://github.com/satijalab/azimuth/blob/243ee5db80fcbffa3452c944254a325a3da2ef9e/R/azimuth.R#L31), and we [previously didn't override the `NULL` default](ttps://github.com/maud-p/OpenScPCA-analysis/blob/0ed2e63923cc2c902954dad5acd6358c37b4c99c/analyses/cell-type-wilms-tumor-06/notebook_template/02a_label-transfer_fetal_full_reference_Cao.Rmd#L105).

I'm ~sure the assay should be RNA, but I'd want to run it through to see if specify "RNA" changes the results. Do you think I should do that?

Reply:
> I'm ~sure the assay should be RNA, but I'd want to run it through to see if specifying* "RNA" changes the results. Do you think I should do that?

I too am ~sure, so I say go for it; it will bring us greater understanding at the very least!